src/couch_mrview/src/couch_mrview_http.erl

```diff
@@ -486,8 +491,43 @@ parse_params(Props, Keys, #mrargs{} = Args0, Options) ->
                 Args0;
             _ ->
                 % group_level set to undefined to detect if explicitly set by user
-                Args0#mrargs{keys = Keys, group = undefined, group_level = undefined}
+                case Keys of
+                    [_] ->
+                        Args0#mrargs{group = undefined, group_level = undefined};
+                    _ ->
+                        Args0#mrargs{keys = Keys, group = undefined, group_level = undefined}
+                end
         end,
+
+    PropsKeys = proplists:get_keys(Props),
+    case config:get_boolean("chttpd", "enable_key_exclusivity", ?DEFAULT_ENABLE_KEY_EXCLUSIVITY) of
+        true ->
+            HasKey = lists:member("key", PropsKeys),
+            HasKeys = lists:member("keys", PropsKeys),
+            HasStartEndKey =
+                lists:foldl(
+                    fun(X, Result) ->
+                        lists:member(X, ["start_key", "startkey", "end_key", "endkey"]) orelse
+                            Result
+                    end,
+                    false,
+                    PropsKeys
+                ),
+            KeyIncompatible =
+                (HasKey andalso HasKeys) orelse
+                    (HasKey andalso HasStartEndKey) orelse
+                    (HasKeys andalso HasStartEndKey),
+
+            case KeyIncompatible of
+                true ->
+                    throw(?ERROR_KEY_INCOMPATIBLE);
+                _ ->
+                    ok
+            end;
+        false ->
+            ok
```

Comment:
This might look better a separate function.

Can use sets to check for membership. It might be simpler than `lists:foldl/3`?

```erlang
PropSet = sets:from_list(Props),
HasKey = sets:is_element("key", PropSet),
HasKeys = sets:is_element("keys", PropSet),
StartEndSet = sets:from_list(["start_key", "startkey", "end_key", "endkey"]),
HasStartEndKey = not sets:is_empty(sets:intersection(StartEndSet, PropSet)).
```

What do you think?